release-22.2: release-22.1: backupccl: reintroduce previously offline tables with manifest.DescriptorChanges

msbutler opened the pull request:
Backport 1/2 commits from #88488.

/cc @cockroachdb/release

---

Backport 1/1 commits from #87312.

/cc @cockroachdb/release

---

getReintroducedSpans finds all tables included in the current and previous
backup that may have undergone a non-mvcc operation. The current backup will
then back up these tables' spans from ts = 0, as the previous backup may have
missed certain non-mvcc operations (e.g. ClearRange, non-mvcc AddSSTable).

To find these tables, getReintroducedSpans must find all tables covered in the
previous backup that were in the offline state at previous backup start time.

This function assumed that the previous backup's
manifest.Descriptors field would contain all tables covered in the previous
backup; however, while investigating https://github.com/cockroachdb/cockroach/issues/88042, we discovered that this assumption
is not correct in pre 22.2 branches. During revision history backups, a table
with an in-progress import (e.g. offline at backup time) can get backed up and
included in manifest.DescriptorChanges but not in manifest.Descriptors. This
implies getReintroducedSpans missed reintroducing spans from this table,
implying that backup chains have missed backing up some data.

It's worth noting this bug could only affect a backup chain in 22.2 iff the
backup that observed the table go offline was taken on a 22.1
cluster. Because 22.2 backups explicitly back up all offline tables, all tables
in manifest.DescriptorChanges should also be in manifest.Descriptors. In other
words, before this patch, the following sequence could lead to a corrupt backup
chain:
- t0: begin IMPORT on foo, on a pre 22.2 cluster
- t1: conduct a revision history backup of foo
- t2: complete or cancel the import of foo, via a non-mvcc operation
- t3: upgrade the cluster to 22.2
- t4: conduct a revision history incremental backup of foo
     - foo's span will not get re-introduced

This patch fixes getReintroducedSpans to ensure it reintroduces tables included
in manifest.DescriptorChanges whose last revision brought the table offline. In
addition, this patch adds a check to restore planning which causes the restore
to fail immedidiately if any restore target is missing an introduced span. The
patch tests this check on a corrupt backup taken on v21.2.

Release note(bug fix): fix apart of Tech Advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

Release justification: bug fix

msbutler committed:
backupccl: reintroduce previously offline tables with manifest.DescriptorChanges

getReintroducedSpans finds all tables included in the current and previous
backup that may have undergone a non-mvcc operation. The current backup will
then back up these tables' spans from ts = 0, as the previous backup may have
missed certain non-mvcc operations (e.g. ClearRange, non-mvcc AddSSTable).

To find these tables, getReintroducedSpans must find all tables covered in the
previous backup that were in the offline state at previous backup start time.

This function assumed that the previous backup's
manifest.Descriptors field would contain all tables covered in the previous
backup; however, while investigating #88042, we discovered that this assumption
is not correct in pre 22.2 branches. During revision history backups, a table
with an in-progress import (e.g. offline at backup time) can get backed up and
included in manifest.DescriptorChanges but not in manifest.Descriptors. This
implies getReintroducedSpans missed reintroducing spans from this table,
implying that backup chains have missed backing up some data.

It's worth noting this bug could only affect a backup chain in 22.2 iff the
backup that observed the table go offline was taken on a 22.1
cluster. Because 22.2 backups explicitly back up all offline tables, all tables
in manifest.DescriptorChanges should also be in manifest.Descriptors. In other
words, before this patch, the following sequence could lead to a corrupt backup
chain:
- t0: begin IMPORT on foo, on a pre 22.2 cluster
- t1: conduct a revision history backup of foo
- t2: complete or cancel the import of foo, via a non-mvcc operation
- t3: upgrade the cluster to 22.2
- t4: conduct a revision history incremental backup of foo
     - foo's span will not get re-introduced

This patch fixes getReintroducedSpans to ensure it reintroduces tables included
in manifest.DescriptorChanges whose last revision brought the table offline. In
addition, this patch adds a check to restore planning which causes the restore
to fail immedidiately if any restore target is missing an introduced span. The
patch tests this check on a corrupt backup taken on v21.2.

Release note(bug fix): fix apart of Tech Advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

Release justification: bug fix

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89102)
<!-- Reviewable:end -->

msbutler committed:
backupccl: only check for introduced spans for online tables in restore

This small patch refines the introduced span corruption checker to
ensure that only _online_ tables in backup i, are checked, as oppose to all
tables included in manifest.Descriptors. This is necessary for the 22.2 patch,
since in 22.2, all backups back up offline descriptors.

msbutler commented:
blathers backport 22.2.0

michae2 commented:
I'm getting these errors when trying to build tip of release-22.2 (d02269f2d35fccbaebf5ebaa09ea0b2d2cba2390):

```
  pkg/ccl/backupccl/backup_planning.go:1065:34: undefined: descpb.FromDescriptor
  pkg/ccl/backupccl/targets.go:378:35: undefined: descpb.FromDescriptor
  pkg/ccl/backupccl/targets.go:436:35: undefined: descpb.FromDescriptor
  pkg/ccl/backupccl/targets.go:448:35: undefined: descpb.FromDescriptor
```

It looks like this PR introduced those lines. I guess 22.1 has [FromDescriptor](https://github.com/cockroachdb/cockroach/blob/e438c2f89282e607e0e6ca1d38b2e0a622f94493/pkg/sql/catalog/descpb/descriptor.go#L227-L234) but 22.2 doesn't.

stevendanna commented:
Looks like this didn't pick up, which is why it passed CI: https://github.com/cockroachdb/cockroach/pull/89152
